### DISCUSSION TITLE
tutorials/2d/2d_movement

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#112] [User-contributed notes]
# tutorials/2d/2d_movement

Introduction: Every beginner has been there: "How do I move my character?" Depending on the style of game you're making, you may have special requirements, but in general the movement in most 2D ga...

https://docs.godotengine.org/en/stable/tutorials/2d/2d_movement.html

**Comment by Tikisearching** [comment]
hello.
I am trying to reate movement for my character and i put the same exact code except for the input names in my game and it does not work. Character is looking at the direction of the mouse but does not move at all. can someone please help me? thanks.

**Reply by Tikisearching** [reply]
i fixed it its good now

**Reply by terrencepolack** [reply]
then tell us how you fixed it please

**Reply by tothricsaj** [reply]
Hey, same issue here - I don't see the fix.

**Reply by tothricsaj** [reply]
I think I found the source of the problem.  If we want to spin the element around own axis, we will have to position the sprite to middle of the main axis. I moved  the spriter therefore it was rotated around the main axis.

**Reply by mmmerge-mightandmagic** [reply]
Hello, I had exactly the same problem: only the character rotation worked, but not the movements using the keyboard and mouse. In my case, the problem was simply Godot's display in Windows!

The solution was simply to go to Editor > Editor Settings > Interface > Editor > change the display scale (I lowered it), and everything worked again after doing that.

**Comment by progress-3325** [comment]
i had a problem and then figured it out so imma say it here in case anyone needs it. you have to write the
	var input_direction = Input.get_vector("left", "right", "up", "down")
part in the parenthesis exactly how it is there, because it first reads negative x, then positive, then negative y and positive y.

**Comment by NelsonDerLord** [comment]
I copy pasted the script for the 8-way movement into my project and the debugger tells me 

The parameter "delta" is never used in the function "_physics_process()". If this is intended, prefix it with an underscore: "_delta".

is this a problem on my end?

**Reply by b-turchyn** [reply]
This is a warning, not an error, but no, it's not a problem. 

It's common practice to prefix unused function parameters with an underscore. This is just a convention, though.

`delta` represents the time since the last frame. If we were manually moving the position of our character, then we'd use `delta` so that our movement appears consistent to the user. [`move_and_slide()`](https://docs.godotengine.org/en/stable/classes/class_characterbody2d.html#class-characterbody2d-method-move-and-slide) accounts for delta time though, and so we don't need to use it.

**Comment by ldinhle** [comment]
For some reason, when following this tutorial, I am unable to get anything other than 8-way movement to work. Any ideas on why?

**Reply by ldinhle** [reply]
For example, when I try to do 'click and move' the sprite moves in the opposite direction away from my mouse

**Reply by samuelshumake** [reply]
I think I just figured this out! It looks like the `position` attribute is relative to the CharacterBody2D, so it's originally set to (0,0) instead of wherever the player is on the map.  used `global_position` instead and grabbed the mouse click event with `get_global_mouse_position()`

```gdscript
extends CharacterBody2D

@export var speed = 400

var target

func _ready():
	target = global_position

func _input(event):
	if event is InputEventMouseButton and event.pressed:
		target = get_global_mouse_position()

func _physics_process(delta):
	velocity = global_position.direction_to(target) * speed
	if global_position.distance_to(target) > 10:
		move_and_slide()
```

**Reply by cnkunz** [reply]
^ this just saved me days of tearing my hair out, this documentation is wrong and needs to be updated. I could not figure out why my character would automatically move to (0,0) at runtime. Thank you so much samuelshumake!

**Comment by TheBlahBlahBlah-max** [comment]
My character is facing the opposite direction when I add 	rotation = get_global_mouse_position().angle_to_point(position)
also when I add a camera 2d to the game my character is can't move/ locked around the camera 2d

**Reply by TheBlahBlahBlah-max** [reply]
I'm trying to implement 8 way movement

**Comment by Feyking** [comment]
When trying to perform the 8-way movement, I noticed that if I held down both the left and right buttons and then tried to press up, the character wouldn't move. However, if I pressed down, it did. The same applies to holding both up and down, then trying to press left, which doesn't work, while pressing right does. This has a noticeable effect when trying to quickly strafe from side to side while moving up, and I was curious if there is a solution to this issue
